subsys/logging/log_core.c

```diff
@@ -0,0 +1,412 @@
+/*
+ * Copyright (c) 2018 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <logging/log_msg.h>
+#include "log_list.h"
+#include <logging/log.h>
+#include <logging/log_backend.h>
+#include <logging/log_ctrl.h>
+#include <logging/log_output.h>
+#include <misc/printk.h>
+#include <assert.h>
+#include <stdio.h>
+
+#ifndef CONFIG_LOG_PRINTK_MAX_STRING_LENGTH
+#define CONFIG_LOG_PRINTK_MAX_STRING_LENGTH 1
+#endif
+
+#ifdef CONFIG_LOG_BACKEND_CONSOLE
+#include <logging/log_backend_console.h>
+LOG_BACKEND_CONSOLE_DEFINE(log_backend_console);
+#endif
+
+static struct log_list_t list;
+
+static bool panic_mode;
+static bool initialized;
+
+static u32_t nullstamp_func(void)
+{
+	return 0;
+}
+
+static timestamp_get_t timestamp_func = nullstamp_func;
+
+
+
+static inline void msg_finalize(struct log_msg *p_msg,
+				struct log_msg_ids src_level)
+{
+	p_msg->hdr.ids = src_level;
+	p_msg->hdr.timestamp = timestamp_func();
+	unsigned int key = irq_lock();
+
+	log_list_add_tail(&list, p_msg);
+
+	irq_unlock(key);
+
+	if (panic_mode) {
+		(void)log_process();
+	}
+}
+void log_0(const char *p_str, struct log_msg_ids src_level)
+{
+	struct log_msg *p_msg = log_msg_create_0(p_str);
+
+	if (p_msg == NULL) {
+		return;
+	}
+	msg_finalize(p_msg, src_level);
+}
+
+void log_1(const char *p_str,
+	   u32_t arg0,
+	   struct log_msg_ids src_level)
```

Comment:
Wait, wait, wait.  Wait.  No, seriously, wait...

You have a staggeringly huge framework constructed of extraordinarily clever preprocessor code.  It automatically detects the varargs structure of the logging used at runtime in order to correctly pack the right number of arguments into the highly compressed stream, so that the code just does the right thing without having to write out all the variants by hand.  It probably does other stuff I haven't figured out yet.

But all that work just lands here in a bunch of hand-coded marshalling functions that need to duplicate all the work via separately-compiled (and thus big!) runtime code?

Why wasn't this whole mess just one C varargs function (with a little wrapper to put the format string into the right section, etc...) to begin with?  Smaller, probably faster, MUCH simpler (seriously: it would cut hundreds on hundreds of lines out of the patch!).

I just don't get why you had to go through all the cpp and boilerplate hell just to get here.

Reply:
The goal was to have best performance possible. Vargs will add some additional cycles. I reused approach we had in logger in nRF5 SDK. Back then I was evaluating the cost of using vargs. I will try to get the numbers as the input for discussion.

Reply:
I've tried to change log macro to build an array of arguments and pass that array to generic log_n() function. Log message is created with single function, etc. It takes 74 cycles more (296 vs. 222) to log typical message (string + 1 argument) on nrf52 (cortex m4, 64mhz). That's 30% worse. In my opinion, it's a significant penalty and I would stay with current solution.